File: CSharp/HtmlTest/Base/CloneTest.cs
using Microsoft.XmlDiffPatch;
using System.IO;
using System.Xml;
using XhtmlXsd;

namespace HtmlTest.Base
{
    public abstract class CloneTest : BaseTest
    {
        protected readonly string Ext;

        public CloneTest() : this("html") { }

        public CloneTest(string ext) : base() => Ext = ext;

        protected CloneTest(string ext, string subFolder) : base(subFolder) => Ext = ext;

        protected override void CompareToExpected(
            FileInfo testOutput, FileInfo testExpected, bool ignoreWhitespace = true
        ) {
            bool pass = false;

            if (Platform.IsXml)
            {
                using XmlTextReader
                    output = new XmlTextReader(testOutput.Open(FileMode.Open, FileAccess.Read, FileShare.Read))
                        { XmlResolver = XhtmlEntityResolver.Instance },
                    expected = new XmlTextReader(testExpected.Open(FileMode.Open, FileAccess.Read, FileShare.Read))
                        { XmlResolver = XhtmlEntityResolver.Instance };

                XmlDiff xmldiff = new XmlDiff(
                    XmlDiffOptions.IgnoreNamespaces |
                    XmlDiffOptions.IgnorePrefixes |
                    XmlDiffOptions.IgnoreDtd |
                    XmlDiffOptions.IgnoreXmlDecl
                );

                pass = xmldiff.Compare(expected, output);
            }

            if (!pass)
                base.CompareToExpected(testOutput, testExpected);
        }

        public void DoTest(FileInfo testInput, FileInfo testOutput, FileInfo testExpected, TestOptions options)
        {
            TestWalker walker;
            if (!Platform.IsXml && options.Formatted && options.DocumentOptions.ProvideEol)
                walker = new FormattedCloneTestWalker(Platform, options.DocumentOptions);
            else
                walker = new CloneTestWalker(Platform, options.DocumentOptions);

            var document = Platform.NewDocument(testInput.FullName, options.DocumentOptions);
            walker.Visit(document.DocumentTag);

            if (testOutput.Exists)
                testOutput.Delete();

            walker.Output.Save(testOutput.FullName, document.Encoding);

            CompareToExpected(testOutput, testExpected);
        }

        public void DoTest(string inFile, string outFile, string expectedFile, TestOptions options)
        {
            var testInput = GetTestInput(inFile);
            var testOutput = GetTestOutput(outFile);
            var testExpected = GetTestExpected(expectedFile, testInput, options);
            DoTest(testInput, testOutput, testExpected, options);
        }
        protected void DoTest(TestOptions options)
            => DoTest("test." + Ext, "clone." + Ext, "expected." + Ext, options);

        protected void DoTest(string t, TestOptions options)
            => DoTest("test." + t + "." + Ext, "clone." + t + "." + Ext, "expected." + t + "." + Ext, options);

        protected void DoTest(int t, TestOptions options)
        {
            options.DocumentOptions.ProvideEol = false;
            DoTest(t.ToString(), options);
            options.DocumentOptions.ProvideEol = true;
            DoTest(t.ToString(), options);
        }

        protected void DoTest(int t) => DoTest(t, new TestOptions(Platform));
    }
}

File: Java/src/htmltest/base/CloneTest.java
package htmltest.base;

import static org.junit.Assert.fail;

import java.io.File;

import htmlwalker.exception.HtmlWalkerException;
import htmlwalker.platform.WalkerDocument;

public abstract class CloneTest extends BaseTest
{
    private final static String OUTPATH = "Test";
	protected final String ext;
	
	protected abstract boolean isXml();

	@Override
	protected String outFolder()  { return OUTPATH; }
    
	public CloneTest() { this("html");	}
	
	public CloneTest(String ext)
	{
		super();
		this.ext = ext;
	}

	public CloneTest(String ext, String subFolder)
	{
		super(subFolder);
		this.ext = ext;
	}
	
	protected void doTest(String inFile, String outFile, String expectedFile, Options options)
	{
	    File testInput = getTestInput(inFile);

	    File testOutput = getTestOutput(outFile);
        if (testOutput.exists())
            testOutput.delete();

	    TestWalker cloner = new TestWalker(platform());
	    cloner.tracer.verbosity = 2;
	    
	    try
	    {
	    	WalkerDocument document = platform().newDocument(testInput.getPath(), options.documentOptions);
	    	cloner.visit(document.documentTag());
	        cloner.output().save(testOutput.getPath());
		}
	    catch (HtmlWalkerException e)
	    {
			fail(e.getMessage());
		}
	
	    if (null != expectedFile)
		    compareToExpected(testOutput, testInput, expectedFile, true);
	}

	protected void doTest(String inFile, String outFile, String expectedFile)
	{
		doTest(inFile, outFile, expectedFile, new Options(platform()));
	}

	protected void doTest(Options options)
    {
	    doTest("test." + this.ext, "clone." + this.ext, "expected." + this.ext, options);
    }

	protected void doTest()
    {
	    doTest("test." + this.ext, "clone." + this.ext, "expected." + this.ext);
    }

    protected void doTest(String t, Options options)
    {
	    doTest("test." + t + "." + this.ext, "clone." + t + "." + this.ext, "expected." + t + "." + this.ext, options);
    }

    protected void doTest(String t)
    {
	    doTest("test." + t + "." + this.ext, "clone." + t + "." + this.ext, "expected." + t + "." + this.ext, new Options(platform()));
    }
	
    protected void doTest(int t, Options options)
    {
    	options.documentOptions.setProvideEol(false);
	    doTest(String.valueOf(t), options);
    	options.documentOptions.setProvideEol(true);
	    doTest(String.valueOf(t), options);
    }
	
    protected void doTest(int t)
    {
	    doTest(t, new Options(platform()));
    }
}
